integration/spark/src/main/scala/org/apache/carbondata/spark/tasks/DictionaryWriterTask.scala

```diff
@@ -63,18 +64,26 @@ class DictionaryWriterTask(valuesBuffer: mutable.HashSet[String],
       if (values.length >= 1) {
         if (model.dictFileExists(columnIndex)) {
           for (value <- values) {
-            if (dictionary.getSurrogateKey(value) ==
-                CarbonCommonConstants.INVALID_SURROGATE_KEY) {
-              val parsedValue = org.apache.carbondata.core.util.DataTypeUtil
+            var newValue = value
+            if (DataType.DECIMAL == model.primDimensions(columnIndex).getDataType) {
+              newValue = org.apache.carbondata.core.util.DataTypeUtil
                 .normalizeColumnValueForItsDataType(value,
                   model.primDimensions(columnIndex))
+            }
+            if (null != newValue && dictionary.getSurrogateKey(newValue) ==
+              CarbonCommonConstants.INVALID_SURROGATE_KEY) {
+              var parsedValue = newValue
+              if (DataType.DECIMAL != model.primDimensions(columnIndex).getDataType) {
+                parsedValue = org.apache.carbondata.core.util.DataTypeUtil
+                  .normalizeColumnValueForItsDataType(value,
+                    model.primDimensions(columnIndex))
+              }
```

Comment:
@Zhangshunyu ...I think you can just move the parsing logic above the dictionary look up logic. This will avoid 2 extra if checks. You can check the sample code below.
if (model.dictFileExists(columnIndex)) {
          for (value <- values) {
            val parsedValue = org.apache.carbondata.core.util.DataTypeUtil
              .normalizeColumnValueForItsDataType(value,
                model.primDimensions(columnIndex))
            if (null != parsedValue && dictionary.getSurrogateKey(value) ==
                CarbonCommonConstants.INVALID_SURROGATE_KEY) {
              if (null != parsedValue) {
                writer.write(parsedValue)
                distinctValues.add(parsedValue)
              }
            }
          }
        }